RecoHI/HiCentralityAlgos/plugins/ClusterCompatibilityProducer.cc

```diff
@@ -0,0 +1,191 @@
+ClusterCompatibilityProducer::~ClusterCompatibilityProducer() {}
+
+void
+ClusterCompatibilityProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  std::auto_ptr<reco::ClusterCompatibility> creco(new reco::ClusterCompatibility());
+
+  // get hold of products from Event
+  edm::Handle<SiPixelRecHitCollection> hRecHits;
+  iEvent.getByToken(inputToken_, hRecHits);
+
+  // get tracker geometry
+  if (hRecHits.isValid()) {
+    edm::ESHandle<TrackerGeometry> trackerHandle;
+    iSetup.get<TrackerDigiGeometryRecord>().get(trackerHandle);
+    const TrackerGeometry *tgeo = trackerHandle.product();
+    const SiPixelRecHitCollection *hits = hRecHits.product();
+
+    // loop over pixel rechits
+    int nPxlHits=0;
+    std::vector<VertexHit> vhits;
+    for(SiPixelRecHitCollection::DataContainer::const_iterator hit = hits->data().begin(),
+          end = hits->data().end(); hit != end; ++hit) {
+      if (!hit->isValid())
+        continue;
+      ++nPxlHits;
+      DetId id(hit->geographicalId());
+      if(id.subdetId() != int(PixelSubdetector::PixelBarrel))
+        continue;
+      const PixelGeomDetUnit *pgdu = static_cast<const PixelGeomDetUnit*>(tgeo->idToDet(id));
+      if (1) {
+        const PixelTopology *pixTopo = &(pgdu->specificTopology());
+        std::vector<SiPixelCluster::Pixel> pixels(hit->cluster()->pixels());
+        bool pixelOnEdge = false;
+        for(std::vector<SiPixelCluster::Pixel>::const_iterator pixel = pixels.begin();
+            pixel != pixels.end(); ++pixel) {
+          int pixelX = pixel->x;
+          int pixelY = pixel->y;
+          if(pixTopo->isItEdgePixelInX(pixelX) || pixTopo->isItEdgePixelInY(pixelY)) {
+            pixelOnEdge = true;
+            break;
+          }
+        }
+        if (pixelOnEdge)
+          continue;
+      }
+
+      LocalPoint lpos = LocalPoint(hit->localPosition().x(),
+                                   hit->localPosition().y(),
+                                   hit->localPosition().z());
+      GlobalPoint gpos = pgdu->toGlobal(lpos);
+      VertexHit vh;
+      vh.z = gpos.z();
+      vh.r = gpos.perp();
+      vh.w = hit->cluster()->sizeY();
+      vhits.push_back(vh);
+    }
+
+    creco->setNValidPixelHits(nPxlHits);
+ 
+    // append cluster compatibility  for each z-position
+    for(double z0 = minZ_; z0 <= maxZ_; z0 += zStep_)
+    { 
+      ContainedHits c = getContainedHits(vhits, z0);
+      creco->append(c.z0, c.nHit, c.chi);
+    }
+
+  }
+  iEvent.put(creco);
+
+}
+
+
+ClusterCompatibilityProducer::ContainedHits ClusterCompatibilityProducer::getContainedHits(const std::vector<VertexHit> &hits, double z0) const
+{
+
+  // Calculate number of hits contained in v-shaped window in cluster y-width vs. z-position.
+  int n = 0;
+  double chi = 0.;
+
+  for(std::vector<VertexHit>::const_iterator hit = hits.begin(); hit!= hits.end(); hit++) {
+    double p = 2 * fabs(hit->z - z0)/hit->r + 0.5; // FIXME <- this comment from the HLT filter, need
```

Comment:
any better understanding to get rid of the FIXME?

also, isn't this more like 1.9 (285/150) instead of 2?